core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java

```diff
@@ -270,6 +276,19 @@ public void close() {
         closeable.close();
     }
 
+    /**
+     * Registers a JVM shutdown hook that closes this {@link ClientFactory} when the current JVM terminates.
+     */
+    public CompletableFuture<Void> closeOnShutdown() {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        Runtime.getRuntime().addShutdownHook(THREAD_FACTORY.newThread(() -> {
+            close();
+            logger.debug("ClientFactory has been closed.");
+            future.join();
+        }));
+        return future;
```

Comment:
Shouldn't we complete the returned `future` when closing is finished?
How about:
```suggestion
        final CompletableFuture<Void> future = new CompletableFuture<>();
        Runtime.getRuntime().addShutdownHook(THREAD_FACTORY.newThread(() -> {
            closeAsync().handle((unused, cause) -> {
                if (cause != null) {
                    logger.warn("Unexpected exception while closing a ClientFactory.", cause);
                    future.completeExceptionally(cause);
                } else {
                    logger.debug("ClientFactory has been closed.");
                    future.complete(null);   
                }
            }
        }));
        return future;
```

Reply:
Yes, we should.